grudge/models/__init__.py

```diff
@@ -49,13 +49,22 @@ def max_characteristic_velocity(self, t, fields, dcoll):
 
     def estimate_rk4_timestep(self, dcoll, t=None, fields=None):
         """Estimate the largest stable timestep for an RK4 method."""
-        from grudge.dt_utils import (dt_non_geometric_factor,
+        from grudge.dt_utils import (dt_non_geometric_factors,
                                      dt_geometric_factors)
+        from meshmode.dof_array import DOFArray
         import grudge.op as op
 
+        actx = dcoll._setup_context
         max_lambda = self.max_characteristic_velocity(t, fields, dcoll)
-        dt_factor = \
-            (dt_non_geometric_factor(dcoll)
-             * op.nodal_min(dcoll, "vol", dt_geometric_factors(dcoll)))
+        # Scale each group array of geometric factors by the corresponding
+        # group non-geometric factor
+        dt_factors = DOFArray(
+            actx,
+            data=(
+                cng * geo_facts
+                for cng, geo_facts in zip(dt_non_geometric_factors(dcoll),
+                                          dt_geometric_factors(dcoll))
+            )
+        )
 
-        return dt_factor * (1 / max_lambda)
+        return op.nodal_min(dcoll, "vol", dt_factors * (1 / max_lambda))
```

Comment:
@MTCam Sorry for the switch-up on this. But could I get your opinion on this? For reference: https://github.com/inducer/grudge/issues/119

Reply:
Is there only one geometric factor per group?  I don't think I understood that before.  In order to get a truly cell-local estimate for dt/cfl, we'll need cell-local geom/nongeom factors.  

I want to explain a little bit what my concern is... In our meshes, we have very tight spacing in the boundary layer, and likely we will refine in recirculation zones where combustion will be initiated.  The flow velocities in these very fine mesh areas should, in general, be much slower.

A common approach is to estimate the global max dt using max wavespeed (e.g. MAX_mesh(|v| + |c|))  and the MIN_mesh(h).  This makes a lot of sense usually (simple geom, simple flow) - but for our cases with complex flow and geometric features - we should be a little smarter, or we end up with an unnecessarily conservative dt (potentially costing us *months* in compute time).  Our PlasCom2 simulations run for weeks-on-end to produce prediction results.  MirgeCom will require more. These seemingly slight differences accumulate a lot over our relatively long runtimes, and have significant impact on the required compute time to do our prediction.

The first stab at doing something better is kind of naive and obvious, but effective. We just do the *cell-local* values for *h*, and the *cell-local* wavespeeds, to get a *cell-local* dt.  The max/min consideration is deferred until the end where we choose the DT by taking a min over cells, or the max CFL by maxing over cells.  MIRGE-Com needs the back-end to support this approach.

All of that said, ANY estimate of the sim cfl or the max dt is going to work for us.  It is not something that ever needs to be precise. The user will take any numbers coming out of this as a *guide* only; the actual stability analysis for any given simulation is an empirical operation.

Reply:
> Is there only one geometric factor per group? I don't think I understood that before. In order to get a truly cell-local estimate for dt/cfl, we'll need cell-local geom/nongeom factors.

Nope, there is one _non geometric_ factor per group. This factor depends on the type of reference cell (triangle, tetrahedron, prim, etc) and the polynomial degree. The geometric factor is an actual measure of the cell diameter, and so will be different for each cell in the mesh (unless you have an evenly spaced, regular mesh).

What this PR does is that it will return all non geometric factors (by group) to you rather than take the minimum over all groups. This line here:
```python
dt_factors = DOFArray(
            actx,
            data=tuple(
                cng * geo_facts
                for cng, geo_facts in zip(dt_non_geometric_factors(dcoll),
                                          dt_geometric_factors(dcoll))
            )
        )
```
produces a dof array with the CFL grid factors at each nodal point (since `dt_geometric_factors` is a dof array of size `(nelem, ndofs)`. The difference here is that the scaling by the non-geometric factor will be different across all groups now (if you have different mesh elements/poly order).

Reply:
Got it @thomasgibson. Thanks for that explanation!

Reply:
I think this: https://github.com/inducer/grudge/pull/121/commits/a3659999b0e5dcfa57de21c1090e4a706b675629 cleans things up a bit and adds a more convenient way to get this information